aws/certdetail.go

```diff
@@ -0,0 +1,189 @@
+package aws
+
+import (
+	"fmt"
+	"log"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws/client"
+)
+
+const (
+	// minimal timeperiod for the NotAfter attribute of a Cert to be in the future
+	minimalCertValidityPeriod = 7 * 24 * time.Hour
+	// used as wildcard char in Cert Hostname/AltName matches
+	glob = "*"
+)
+
+// CertDetail is the business object for Certificates
+type CertDetail struct {
+	Arn       string
+	AltNames  []string
+	NotBefore time.Time
+	NotAfter  time.Time
+}
+
+type certificateCache struct {
+	sync.Mutex
+	certDetails []*CertDetail
+	iamClient   *iamClient
+	acmClient   *acmClient
+}
+
+func newCertCache(p client.ConfigProvider) *certificateCache {
+	return &certificateCache{
+		acmClient:   newACMClient(p),
+		iamClient:   newIAMClient(p),
+		certDetails: make([]*CertDetail, 0),
+	}
+}
+
+// updateCertCache will only update the current certificateCache if
+// all calls to AWS API were successfull and not rateLimited for
+// example. In case it could not update the certificateCache it will
+// return the orginal error.
+func (cc *certificateCache) updateCertCache() error {
+	iamList := make([]*CertDetail, 0)
+	acmList := make([]*CertDetail, 0)
+	var iamError *error
+	var acmError *error
+	var wg sync.WaitGroup
+	wg.Add(1)
+	cc.iamClient.updateCerts(iamList, iamError, &wg)
+	wg.Add(1)
+	cc.acmClient.updateCerts(acmList, acmError, &wg)
+	wg.Wait()
+
+	if iamError != nil {
+		return fmt.Errorf("Error in the IAM worker: %s", *iamError)
+	}
+	if acmError != nil {
+		return fmt.Errorf("Error in the ACM worker: %s", *acmError)
+	}
+	cc.Lock()
+	cc.certDetails = append(iamList, acmList...)
+	cc.Unlock()
+	return nil
+}
+
+func (cc *certificateCache) backgroundCertCacheUpdate(certUpdateInterval time.Duration) {
+	go func() {
+		for {
+			time.Sleep(certUpdateInterval)
+			if err := cc.updateCertCache(); err != nil {
+				log.Printf("Cert cache update failed, caused by: %v", err)
+			} else {
+				log.Println("Successfully updated cert cache")
+			}
+		}
+	}()
+}
+
+// GetCachedCerts returns a copy of the cached acm CertificateDetail slice
+// https://docs.aws.amazon.com/acm/latest/APIReference/API_ListCertificates.html#API_ListCertificates_RequestSyntax
+// filtered by CertificateStatuses
+// https://docs.aws.amazon.com/acm/latest/APIReference/API_ListCertificates.html#API_ListCertificates_RequestSyntax
+func (cc *certificateCache) GetCachedCerts() []*CertDetail {
+	cc.Lock()
+	copy := cc.certDetails[:]
+	cc.Unlock()
+
+	return copy
+}
+
+// FindBestMatchingCertificate get all ACM certificates and use a suffix
+// search best match operation in order to find the best matching
+// certificate ARN.
+//
+// We don't need to validate the Revocation here, because we only pull
+// ISSUED certificates.
+func FindBestMatchingCertificate(certs []*CertDetail, hostname string) (*CertDetail, error) {
+	candidate := &CertDetail{}
```

Comment:
Make this `var candidate *CertDetail` and drop the check for the longestMatch below.
Also just realized that longestMatch is flawed: `a.domain.com` is more specific than `*.domain.com` although both have the same length.